traitsui/qt4/datetime_editor.py

```diff
@@ -15,60 +15,79 @@
 """
 
 
-import datetime
-
-from pyface.qt import QtGui
+from pyface.qt import QtGui, qt_api
 from pyface.qt.QtCore import QDateTime
-from traits.api import Instance
+from traits.api import Datetime, on_trait_change
 
 from .editor import Editor
 from .editor_factory import ReadonlyEditor as BaseReadonlyEditor
 
 
+
 class SimpleEditor(Editor):
     """ Simple Traits UI time editor that wraps QDateTimeEdit.
     """
 
     #: the earliest datetime allowed by the editor
-    minimum_datetime = Instance(datetime.datetime)
+    minimum_datetime = Datetime
 
     #: the latest datetime allowed by the editor
-    maximum_datetime = Instance(datetime.datetime)
+    maximum_datetime = Datetime
 
     def init(self, parent):
         """ Finishes initializing the editor by creating the underlying toolkit
             widget.
         """
-        self.control = QtGui.QDateTimeEdit()
+        # set min and max early, don't wait for editor sync
         self.minimum_datetime = self.factory.minimum_datetime
         self.maximum_datetime = self.factory.maximum_datetime
-        self.update_date_range()
+
+        self.control = QtGui.QDateTimeEdit()
         self.control.dateTimeChanged.connect(self.update_object)
+        self.update_date_range()
```

Comment:
Is this the same as the following?:
```
    def init(self, parent):
        self.control = QtGui.QDateTimeEdit()
        self.control.dateTimeChanged.connect(self.update_object)
        self.minimum_datetime = self.factory.minimum_datetime
        self.maximum_datetime = self.factory.maximum_datetime

    # remove update_date_range

    @on_trait_change('minimum_datetime')
    def update_minimum_datetime(self):
        ...

    @on_trait_change('maximum_datetime')
    def update_maximum_datetime(self):
        ...
```

Somehow this fits my brain better. Currently `update_date_range` is called three times if `minimum_datetime` and `maximum_datetime` are not None, the first two calls for setting `self.minimum_datetime` and `self.maximum_datetime` actually do nothing because `control` is None.

Reply:
Also, Qt will adjust the minimum/maximum datetime in order to maintain the validity of the range.
E.g. if the maximum datetime is actually earlier than the minimum datetime.

From the documentation on setting maximumDateTime:
> When setting this property, the minimumDateTime is adjusted, if necessary, to ensure that the range remains valid. Otherwise, changing this property preserves the minimumDateTime property.

If that happens, it is possible for the minimum/maximum datetime on the editor to become out-of-sync with the corresponding values on the Qt widget. In any case, this is worth a separate issue to investigate and fix.